tools/wake-format/emitter.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include "emitter.h"
+
+#include <cassert>
+
+#include "parser/parser.h"
+
+Emitter::nest_t Emitter::nest(ctx_t ctx) { return nest_t(&ctx); }
+
+wcl::rope Emitter::newline(ctx_t ctx) {
+  wcl::rope_builder builder;
+
+  builder.append("\n");
+  for (int i = 0; i < ctx.nest_level; i++) {
+    builder.append(space(ctx, space_per_indent));
+  }
+
+  return std::move(builder).build();
+}
+
+wcl::rope Emitter::space(ctx_t ctx, uint8_t count) {
+  wcl::rope_builder builder;
+  for (uint8_t i = 0; i < count; i++) {
+    builder.append(" ");
+  }
+  return std::move(builder).build();
+}
+
+wcl::optional<wcl::rope> Emitter::flat(ctx_t ctx, CSTElement node) {
+  return walk_top(ctx.flat(), node);
+}
```

Comment:
I'm wondering now if we're going to have several things like top, expression, etc... that we need to memorize for separately. No request here just a thought. It might mean that we don't have a single `flat` method just we always use `ctx.flat()` to request the flat version of something.

This would be needed if a CST element needs to be contextually formatted a certain way that isn't apt for encoding in the context. I can't think of any examples now but worth keeping in mind that we have that out if we need it.

Reply:
Yeah exactly. I think we probably only need the things that dispatch to memoize. Like `expression` will memoize but that might call `match` which calls `case` which then calls `expression` again and I think only expression needs to memoize there. The cool thing is that memoizing is fairly simple since the CSTElements contain an index and the context is easily hashable/orderable and ropes are very cheap to copy construct and move.